### PR TITLE
add comments and minor fixes to Conv2D.

### PR DESCRIPTION
**Pull request by t-k-**

### CODE REVIEW

**Comment by t-k-** [comment]
Sure, no problem. 

BTW. I find your implementation for conv layer is a little hard to follow, and I have read that the original caffe and also darknet (Yolo) code, they both use [im2col implementation](https://github.com/Yangqing/caffe/wiki/Convolution-in-Caffe:-a-memo) which may worth to have, due to the simplicity and easy to port to cuda using GEMM. Although the memory consumption is huge, there is techniques to balance the memory-performance tradeoff using "set".

**Comment by borgwang** [comment]
Correct me if I’m wrong. The basic idea is the same as img2col. 

> The trick is to just lay out all the local patches, and organize them to a (W*H, K*K*D) matrix. In Matlab, this is usually known as an im2col operation. After that, consider the filters being a (M, K*K*D) matrix too, the convolution naturally gets reduced to a matrix multiplication (Gemm in BLAS) problem.

**Comment by t-k-** [comment]
@borgwang You are right, the high level idea is the same, but implementation wise, there is some version that looks easier to me (also in terms of translating to C code). For example the im2col part below. 
```py
def im2col(img, ksize, stride=1):
    N, H, W, C = img.shape
    out_h = (H - ksize) // stride + 1
    out_w = (W - ksize) // stride + 1
    col = np.empty((N * out_h * out_w, ksize * ksize * C))
    outsize = out_w * out_h
    for y in range(out_h):
        y_min = y * stride
        y_max = y_min + ksize
        y_start = y * out_w
        for x in range(out_w):
            x_min = x * stride
            x_max = x_min + ksize
            col[y_start+x::outsize, :] = img[:, y_min:y_max, x_min:x_max, :].reshape(N, -1)
    return col
```
And also the conv function
```
def conv(X, W, stride=1, padding='same'):
    FN, ksize, ksize, C = W.shape
    if padding == 'same':
	p = ksize // 2
	X = np.pad(X, ((0, 0), (p, p), (p, p), (0, 0)), 'constant')
    N, H, W, C = X.shape
    col = im2col(X, ksize, stride)
    z = np.dot(col, W.reshape(W.shape[0], -1).transpose())
    z = z.reshape(N, z.shape[0] / N, -1)
    out_h = (H - ksize) // stride + 1
    return z.reshape(N, out_h, -1 , FN)
```
Although I am not sure it is more efficient in Python and I'm not in front of a computer to try it out. I am going to be busy with other things in the next couple of days, otherwise I would love to test it.

**Comment by t-k-** [comment]
@borgwang I have tried to mimic this approach and created a newer version of conv2d layer which I feel easier to understand in pull request #21 .

Let me know what do you think. I would love to take some time revising it if you consider to change to this version.
